### ISSUE TITLE
Blank bindings file for Objective-C headers via ffigen

### ISSUE THREAD

**Issue by leighajarett** [dart-lang/ffigen#470]
I am trying to use ffigen to generate the dart bindings for some of Apple's APIs. However, every one I have tried (with the exception of the examples in the documentation) have resulted in a blank bindings file. 

I am on a Mac, and I am am using dart version 2.19

Example 1
[Core Motions API](https://developer.apple.com/documentation/coremotion)

- Pubspec 

```
ffigen:
    name: IntentsLibrary
    language: objc
    output: "foundation_bindings.dart"
    exclude-all-by-default: true
    objc-interfaces:
      include:
        - "CMMotionManager"
    headers:
      entry-points:
        - "/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk/System/Library/Frameworks/CoreMotion.framework/Headers/CMMotionManager.h"
```

- Output
```
Building package executable... (1.4s)
Built ffigen:ffigen.
Running in Directory: '/Users/leighaj/dart/ffi_example/ffi_example'
[SEVERE] : Objective C support is EXPERIMENTAL. The API may change in a breaking way without notice.
[WARNING]: Prefer adding Key '[description]' to your config.
Input Headers: [/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk/System/Library/Frameworks/CoreMotion.framework/Headers/CMMotionManager.h]
[SEVERE] : Header /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk/System/Library/Frameworks/CoreMotion.framework/Headers/CMMotionManager.h: Total errors/warnings: 1.
[SEVERE] :     /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk/System/Library/Frameworks/CoreMotion.framework/Headers/CMMotionManager.h:77:117: warning: 'visibility' attribute ignored [Semantic Issue]
Finished, Bindings generated in /Users/leighaj/dart/ffi_example/ffi_example/foundation_bindings.dart
```

Example 2
[HealthKit](https://developer.apple.com/documentation/healthkit/hkhealthstore?language=objc)

- Pubspec
```
ffigen:
    name: IntentsLibrary
    language: objc
    output: "foundation_bindings.dart"
    exclude-all-by-default: true
    objc-interfaces:
      include:
        - "HKHealthStore"
    headers:
      entry-points:
        - "/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk/System/Library/Frameworks/HealthKit.framework/Headers/HKHealthStore.h"
```

- Output
```
Building package executable... (1.5s)
Built ffigen:ffigen.
Running in Directory: '/Users/leighaj/dart/ffi_example/ffi_example'
[SEVERE] : Objective C support is EXPERIMENTAL. The API may change in a breaking way without notice.
[WARNING]: Prefer adding Key '[description]' to your config.
Input Headers: [/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk/System/Library/Frameworks/HealthKit.framework/Headers/HKHealthStore.h, /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk/System/Library/Frameworks/HealthKit.framework/Headers/HKDefines.h]
[SEVERE] : Header /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk/System/Library/Frameworks/HealthKit.framework/Headers/HKHealthStore.h: Total errors/warnings: 1.
[SEVERE] :     /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk/System/Library/Frameworks/HealthKit.framework/Headers/HKHealthStore.h:8:9: fatal error: 'HealthKit/HKDefines.h' file not found [Lexical or Preprocessor Issue]
Finished, Bindings generated in /Users/leighaj/dart/ffi_example/ffi_example/foundation_bindings.dart
```

- Note that the HKDefines.h file is in the same directory, with the path: 
```
/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk/System/Library/Frameworks/HealthKit.framework/Headers/HKDefines.h
```
       

Let me know if these should be separated into different issues! :) 

Are there specific things we should be considering when identifying potential APIs for FFI with Objective-C?

**Comment by liamappelbe** [comment]
@leighajarett It's interesting that these are all clang compile errors. This means that either the clang version is wrong, or there's some missing compiler flags. What's your clang version (`clang --version`)? 

I can't reproduce the CMMotionManager.h issue. I get a 60k line generated file.

I can reproduce the HKDefines error. HKDefines.h is in the same directory as HKHealthStore.h, but it's being included using a relative path, `HealthKit/HKDefines.h`. I can fix this issue by adding this to the config:

```yaml
    compiler-opts:
      - '-F/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk/System/Library/Frameworks'
```

But then I get a ton of clang compile errors that say it's missing `uint8_t`, `uint64_t` etc. @dcharkes Have you seen this issue before, where the basic int types are missing?

**Comment by dcharkes** [comment]
That vaguely rings a bell. Somewhere in `/Applications/Xcode.app/Contents` should be files with the standard C definitions. From old notes from mine: `clang -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/ ...` does that solve the issue?

**Comment by liamappelbe** [comment]
Thanks Daco. I made progress by adding more compile flags in the config:

```yaml
      - '-I/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk/usr/include'
      - '-arch arm64'
      - '-Wno-nullability-completeness'
```

But it turns out those warnings were a red herring. There's a bug in `shouldIncludeRootCursor` where the source file filtering is too strict.

**Comment by leighajarett** [comment]
@liamappelbe my clang version is 13.1.6

**Comment by liamappelbe** [comment]
@leighajarett What's the full version string? Does it say "Homebrew" or "Apple"? Either should work, but Apple's might be more reliable for iOS stuff.

To summarize, I think there's 2 issues here:
1. The main issue is the root cursor filtering bug, so we'll see what you get after #471 lands.
2. You'll probably need to set the `-F` compiler option I mentioned [above](https://github.com/dart-lang/ffigen/issues/470#issuecomment-1268715017) to fix any header not found issues

From my testing, the other warnings don't matter too much, though you can add more of those compiler options to quiet them.

**Comment by leighajarett** [comment]
Apple clang version 13.1.6 (clang-1316.0.21.2.5)

Thanks for your help with this! I want this demo to be applicable for iOS developers, so in general, when they are using this feature should they always use the compiler option you mentioned?

**Comment by liamappelbe** [comment]
Yeah, I think they'll usually need that -F option. It tells the compiler where to find the frameworks.

**Comment by liamappelbe** [comment]
@dcharkes I'm OOO for a few days. Can you publish v7 when you get a chance to unblock Leigha?

**Comment by dcharkes** [comment]
I have published version v7.0.0.

@leighajarett let us know if you have more questions.
